### PR TITLE
draft chainlink oracle

### PR DESCRIPTION
**Pull request by wminshew**
[spec](https://www.notion.so/splits/spec-chainlink-oracle-c97fbc5b01c74788a5a9dc7801ced5b5)

outstanding items:
- [ ] consider multihop
- [ ] tests
- [ ] L2 support
- [ ] default chainlink oracle pairs / feeds [ for each network ]
- [ ] default chainlink oracle script [ for each network ]

### CODE REVIEW

**Comment by wminshew** [comment]
as a note, when it comes to tokenOverrides we generally need to use something that isn't 0x0 (which we normally use for ETH) here i think (can play around w the design to see if we can / want to avoid this, but CL's eth denomination is [0xEeee](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/Denominations.sol#L6) anyway)

as an aside, perhaps this is a good reason to not use 0x0 for ETH in v2

File: src/ChainlinkOracleImpl.sol [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+
+    /// default time until CL updates become stale
+    /// 4 bytes
+    uint32 internal $defaultStaleAfter;
+
+    /// slot 1 - 0 bytes free
+
+    /// overrides for tokens
+    /// 32 bytes
+    mapping(address => address) internal $_tokenOverrides;
```
Comment (mihoward21) [inline src/ChainlinkOracleImpl.sol:103]: what exactly is this for? you can just arbitrarily have any token overridden to use a different token?

File: src/ChainlinkOracleImpl.sol [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        if (invertFeed) {
+            quoteAmount = uint256(quoteParams_.baseAmount) * 10 ** uint256(answerDecimals) / uint256(answer);
+        } else {
+            quoteAmount = uint256(quoteParams_.baseAmount) * uint256(answer) / 10 ** uint256(answerDecimals);
+        }
+        return _adjustQuoteDecimals(quoteAmount, quoteParams_);
+    }
+
+    /// adjust quoteAmount_ based on decimals of base & quote
+    function _adjustQuoteDecimals(uint256 quoteAmount_, QuoteParams calldata quoteParams_)
```
Comment (mihoward21) [inline src/ChainlinkOracleImpl.sol:286]: i should probably read up on chainlink to get a better understanding of it. think i'm mostly following the code here, but just wanted to double check that this `adjustQuoteDecimals` function doesn't need to take into account the inverted state.

Having a little trouble following the math, probably need to run through a sample so i can wrap my head around it.

File: src/ChainlinkOracleImpl.sol [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        returns (uint256)
+    {
+        uint8 baseDecimals = quoteParams_.quotePair.base._decimals();
+        uint8 quoteDecimals = quoteParams_.quotePair.quote._decimals();
+
+        int256 decimalAdjustment = int256(uint256(quoteDecimals)) - int256(uint256(baseDecimals));
+        if (decimalAdjustment > 0) {
+            return quoteAmount_ * 10 ** uint256(decimalAdjustment);
+        }
+        return quoteAmount_ / 10 ** uint256(-decimalAdjustment);
```
Comment (mihoward21) [inline src/ChainlinkOracleImpl.sol:298]: just double checking this works ok if it's 0 (not sure if `-0` will cause an issue).

File: src/libraries/ChainlinkPairDetails.sol [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        function (address) internal view returns (address) convert_,
+        ChainlinkOracleImpl.SetPairDetailParams calldata params_
+    ) internal {
+        ConvertedQuotePair memory cqp = params_.quotePair._convert(convert_);
+        SortedConvertedQuotePair memory scqp = cqp._sort();
+        if (cqp.cBase == scqp.cToken0) {
+            self[scqp.cToken0][scqp.cToken1] = params_.pairDetail;
+        } else {
+            ChainlinkOracleImpl.PairDetail memory pairDetail = params_.pairDetail;
+            pairDetail.inverted = !pairDetail.inverted;
```
Comment (mihoward21) [inline src/libraries/ChainlinkPairDetails.sol:39]: hmm. something feels funny about this inverted property. i'm a little confused why it needs to get stored.

Comment (mihoward21) [reply src/libraries/ChainlinkPairDetails.sol:39]: it's possible to store the wrong boolean value for it, right? there's no way to verify it's correct, just need to trust that the oracle owner set it up correctly?

**Comment by wminshew** [comment]
multihop: use intermediary token whitelist [for most pairs on most chains, only ETH/USDC will be needed as CL topology follows a ~hub-spoke model]

**Pull request closed without merging**